zap/src/output/luau/mod.rs

```diff
@@ -18,69 +15,526 @@ pub trait Output {
+	}
+
+	fn push_range_check(&mut self, range: Range, val: &str) {
+		if let Some(exact) = range.exact() {
+			self.push_line(&format!("assert({val} == {exact})"));
+		} else {
+			if let Some(min) = range.min() {
+				self.push_line(&format!("assert({val} >= {min})"));
+			}
+
+			if let Some(max) = range.max() {
+				self.push_line(&format!("assert({val} <= {max})"));
+			}
 		}
+	}
+
+	fn push_ser(&mut self, from: &str, ty: &Ty<'_>, checks: bool) {
+		self.push_line_indent("do");
+
+		match ty {
+			Ty::Num(numty, range) => {
+				if checks {
+					self.push_range_check(*range, from);
+				}
+
+				self.push_line(&format!("alloc({})", numty.size()));
+				self.push_line(&format!("buffer.write{numty}(outgoing_buff, outgoing_apos, {from})"));
+			}
+
+			Ty::Str(len) => {
+				if let Some(exact) = len.exact() {
+					if checks {
+						self.push_line(&format!("assert(#{from} == {exact})"));
+					}
+
+					self.push_line(&format!("alloc({exact})"));
+					self.push_line(&format!(
+						"buffer.writeu16(outgoing_buff, outgoing_apos, {from}, {exact})"
+					));
+				} else {
+					self.push_line(&format!("local len = #{from}"));
+
+					if checks {
+						self.push_range_check(*len, "len");
+					}
+
+					self.push_line("alloc(len)");
+					self.push_line("buffer.writeu16(outgoing_buff, outgoing_apos, len)");
+					self.push_line(&format!(
+						"buffer.writestring(outgoing_buff, outgoing_apos, {from}, len)"
+					));
+				}
+			}
+
+			Ty::Buf(len) => {
+				if let Some(exact) = len.exact() {
+					if checks {
+						self.push_line(&format!("assert(#{from} == {exact})"));
+					}
 
-		match &stmt {
-			Stmt::Local(name, expr) => {
-				if let Some(expr) = expr {
-					self.push_line(&format!("local {name} = {expr}"));
+					self.push_line(&format!("alloc({exact})"));
+					self.push_line(&format!(
+						"buffer.copy(outgoing_buff, outgoing_apos, {from}, 0, {exact})"
+					));
 				} else {
-					self.push_line(&format!("local {name}"));
+					self.push_line(&format!("local len = #{from}"));
+
+					if checks {
+						self.push_range_check(*len, "len");
+					}
+
+					self.push_line("alloc(len)");
+					self.push_line(&format!("buffer.copy(outgoing_buff, outgoing_apos, {from}, 0, len)"));
 				}
 			}
-			Stmt::LocalTuple(var, expr) => {
-				let items = var.join(", ");
 
-				if let Some(expr) = expr {
-					self.push_line(&format!("local {items} = {expr}"));
+			Ty::Arr(ty, len) => {
+				if let Some(exact) = len.exact() {
+					if checks {
+						self.push_line(&format!("assert(#{from} == {exact})"));
+					}
+
+					self.push_line_indent(&format!("for i = 1, {exact} do"));
+					self.push_line(&format!("local value = {from}[i]"));
+					self.push_ser("value", ty, checks);
+					self.push_dedent_line("end");
 				} else {
-					self.push_line(&format!("local {items}"));
+					self.push_line(&format!("local len = #{from}"));
+
+					if checks {
+						self.push_range_check(*len, "len");
+					}
+
+					self.push_line("alloc(len)");
```

Comment:
should alloc for the u16 and ser allocs the rest

```suggestion
					self.push_line("alloc(2)");
```